src/Workspaces/Core/Portable/FlowAnalysis/SymbolUsageAnalysis/SymbolUsageAnalysis.DataFlowAnalyzer.FlowGraphAnalysisData.cs

```diff
@@ -0,0 +1,499 @@
+                        analyzeLocalFunctionOrLambdaInvocation,
+                        reachingDelegateCreationTargets: PooledDictionary<IOperation, PooledHashSet<IOperation>>.GetInstance(),
+                        localFunctionTargetsToAccessingCfgMap: PooledDictionary<IMethodSymbol, ControlFlowGraph>.GetInstance(),
+                        lambdaTargetsToAccessingCfgMap: PooledDictionary<IFlowAnonymousFunctionOperation, ControlFlowGraph>.GetInstance());
+                }
+
+                public static FlowGraphAnalysisData Create(
+                    ControlFlowGraph cfg,
+                    IMethodSymbol lambdaOrLocalFunction,
+                    FlowGraphAnalysisData parentAnalysisData)
+                {
+                    Debug.Assert(cfg.Parent != null);
+                    Debug.Assert(lambdaOrLocalFunction.IsAnonymousFunction() || lambdaOrLocalFunction.IsLocalFunction());
+                    Debug.Assert(parentAnalysisData != null);
+
+                    var parameters = lambdaOrLocalFunction.GetParameters();
+                    return new FlowGraphAnalysisData(
+                        cfg,
+                        parameters,
+                        analysisDataByBasicBlockMap: CreateAnalysisDataByBasicBlockMap(cfg),
+                        symbolsWriteMap: UpdateSymbolsWriteMap(parentAnalysisData.SymbolsWriteBuilder, parameters),
+                        symbolsRead: parentAnalysisData.SymbolsReadBuilder,
+                        lambdaOrLocalFunctionsBeingAnalyzed: parentAnalysisData.LambdaOrLocalFunctionsBeingAnalyzed,
+                        analyzeLocalFunctionOrLambdaInvocation: parentAnalysisData._analyzeLocalFunctionOrLambdaInvocation,
+                        reachingDelegateCreationTargets: parentAnalysisData._reachingDelegateCreationTargets,
+                        localFunctionTargetsToAccessingCfgMap: parentAnalysisData._localFunctionTargetsToAccessingCfgMap,
+                        lambdaTargetsToAccessingCfgMap: parentAnalysisData._lambdaTargetsToAccessingCfgMap);
+                }
+
+                private static PooledDictionary<BasicBlock, BasicBlockAnalysisData> CreateAnalysisDataByBasicBlockMap(
+                    ControlFlowGraph cfg)
+                {
+                    var builder = PooledDictionary<BasicBlock, BasicBlockAnalysisData>.GetInstance();
+                    foreach (var block in cfg.Blocks)
+                    {
+                        builder.Add(block, null);
+                    }
+
+                    return builder;
+                }
+
+                public ControlFlowGraph ControlFlowGraph { get; }
+
+                /// <summary>
+                /// Flow captures for l-value or address captures.
+                /// </summary>
+                public ImmutableHashSet<CaptureId> LValueFlowCapturesInGraph { get; }
+
+                public BasicBlockAnalysisData GetCurrentBlockAnalysisData(BasicBlock basicBlock)
```

Comment:
>Current [](start = 49, length = 7)

1. maybe Current is not necessary
2. actually, the pattern allows exceptions if key not found. maybe it is better to avoid

Reply:
Thanks. Fixed 1. Regarding 2, `CreateAnalysisDataByBasicBlockMap` invoked in the constructor guarantees there is always an entry in the map for each basic block.

---
In reply to: [234060411](https://github.com/dotnet/roslyn/pull/31025#discussion_r234060411) [](ancestors = 234060411)